src/test/java/hudson/plugins/git/extensions/impl/PruneStaleTagTest.java

```diff
@@ -0,0 +1,237 @@
+import hudson.Functions;
+import hudson.model.Run;
+import hudson.model.TaskListener;
+import hudson.plugins.git.GitSCM;
+import hudson.util.LogTaskListener;
+import java.io.File;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import nl.jqno.equalsverifier.EqualsVerifier;
+import org.apache.commons.io.FileUtils;
+import org.jenkinsci.plugins.gitclient.GitClient;
+import org.jenkinsci.plugins.gitclient.TestCliGitAPIImpl;
+import org.jenkinsci.plugins.gitclient.TestJGitAPIImpl;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import static org.mockito.Mockito.*;
+
+public class PruneStaleTagTest {
+
+    @Rule
+    public TemporaryFolder fileRule = new TemporaryFolder();
+
+    private TaskListener listener;
+    private Run<?, ?> run;
+
+    @Before
+    public void setup() throws Exception {
+        listener = new LogTaskListener(Logger.getLogger("prune tags"), Level.FINEST);
+
+        run = mock(Run.class);
+        when(run.getEnvironment(listener)).thenReturn(new EnvVars());
+    }
+
+    @Test
+    public void equalsContract() {
+        EqualsVerifier.forClass(PruneStaleTag.class)
+                .usingGetClass()
+                .verify();
+    }
+
+    /*
+     * Local tag       | Remote tag
+     * ---------------------------------
+     * exists - differ | exists - differ
+     */
+    @Test
+    public void verify_local_tag_is_pruned_if_different_than_on_remote() throws Exception {
+        File remoteRepo = fileRule.newFolder("remote");
+
+        // create a remote repository with one tag
+        GitClient remoteClient = initRepository(listener, remoteRepo);
+        String tagName = "tag";
+        String tagComment = "tag comment";
+        remoteClient.tag(tagName, tagComment);
+
+        // clone remote repository to workspace
+        GitClient localClient =  cloneRepository(remoteRepo);
+
+        GitSCM scm = new GitSCM(localClient.getRemoteUrl("origin"));
+        PruneStaleTag extension = new PruneStaleTag();
+
+        // get remote SHA1 for the tag
+        String remoteTagHash = remoteClient.getTags().stream().filter(t -> tagName.equals(t.getName())).findFirst().get().getSHA1String();
+
+        FileUtils.touch(new File(localClient.getWorkTree().getRemote(), "localTest"));
+        localClient.add("localTest");
+        localClient.commit("more commits");
+        localClient.deleteTag(tagName);
+        localClient.tag(tagName, tagComment);
+        String localHashTag = localClient.getTags().stream().filter(t -> tagName.equals(t.getName())).findFirst().get().getSHA1String();
+        Assert.assertNotEquals("pre validation failed, local tag must not be the same than remote", remoteTagHash, localHashTag);
+
+        extension.decorateFetchCommand(scm, run, localClient, listener, null);
+        Assert.assertFalse("local tag differ from remote tag and is not pruned", localClient.tagExists(tagName));
```

Comment:
A bit confused by this line: which shouldn't this `tagName` exist?

Reply:
The method comment summarize the use case asked by MarkEWaite.
Local tag points to a different tag than remote.
local tag "1.0.0" -> SHA1(abc123)
remote tag "1.0.0" -> SHA1(123abc)
after the execution I expect the local tag is removed before the job build execution.